Add files via upload

houlinyu opened the pull request:
modified from the function isoquant_matrix() in io.py the reference. file 'gene_Ensembl_to_name.tsv' is used to transfer Ensembl ID to gene name. the outputs can be directly read in by Scanpy (python based downstream) and Seurat (R based downstream)

houlinyu committed:
Add files via upload

modified from the function isoquant_matrix() in io.py
the reference file 'gene_Ensembl_to_name.tsv' is used to transfer Ensembl ID to gene name
the outputs can be directly read in by Scanpy (python based downstream) and Seurat (R based downstream)

jamestwebber commented:
We shouldn't commit the data file to the repo. It's a big file and it'll be sample-specific anyway (e.g. mouse vs human vs other things).

We do have a need for this kind of mapping information, though. Perhaps some more generic code to parse a GTF, or a set of files in `gs://mdl-refs` that we all use.

jamestwebber commented on src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py:
```diff
@@ -0,0 +1,123 @@
+import csv
+from collections import defaultdict
+from pathlib import Path
+from typing import Sequence
+
+import h5py
+import numpy as np
+import scipy.io
+import sparse
+
+import pandas as pd
+import shutil
+import pysam
+import os
+import gzip
+
+import scanpy as sc
```
Please use `black` and `ruff` to format the file (this will include sorting the imports). See [here](https://github.com/MethodsDev/mdl-core/blob/main/code-style.md) for more details (and also the config for this repo).

jamestwebber commented on src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py:
```diff
@@ -0,0 +1,123 @@
+):
+    """
+    Takes the output file from IsoQuant, along with a mapping from read name to
+    barcode+umi. Returns spase matrix, together with barcode and feature lists. Output also include h5ad file which can be directly used for scanpy downstream analysis. The size of the array depends on the size of the barcode and feature lists, if they are provided.
```
this function doesn't appear to return anything, so this should be updated

jamestwebber commented on src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py:
```diff
@@ -0,0 +1,123 @@
+        out_h5ad: Path to h5ad file
+        gene_Ensembl_to_name: a tsv file with first column contaning Ensembl ID and second column containing gene name
+    
+    Opututs:
```
```suggestion
    Outputs:
```

jamestwebber commented on src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py:
```diff
@@ -0,0 +1,123 @@
+    """
+    valid_assignments = set(valid_assignments)
+    
+    samfile = pysam.AlignmentFile(bam_path, "rb")
+    read_to_barcode_umi = {}
+    for read in samfile:
+        read_to_barcode_umi.update({read.query_name:(read.get_tag("CB"),read.get_tag("XM"))})
+
+    rname_to_tx = dict()
+    tx_umi_count = defaultdict(lambda: defaultdict(set))
+
+    with open(isoquant_path) as fh:
+        fh.readline()
+        fh.readline()
+        for r in csv.DictReader(fh, delimiter="\t"):
+            if r["#read_id"] in read_to_barcode_umi:
+                if r["assignment_type"] in valid_assignments:
+                    rname_to_tx[r["#read_id"]] = (r["isoform_id"], r["gene_id"])
+                    bc, umi = read_to_barcode_umi[r["#read_id"]]
+                    tx_umi_count[bc][(r["isoform_id"], r["gene_id"])].add(umi)
+
+    if barcode_list is None:
+        barcode_list = sorted(tx_umi_count)
+
+    barcode_index = {bc: i for i, bc in enumerate(barcode_list)}
+
+    if feature_list is None:
+        feature_list = sorted(set(rname_to_tx.values()))
+
+    feature_index = {tx: i for i, tx in enumerate(feature_list)}
+
+    matrix = sparse.COO.from_iter(
+        (
+            ((barcode_index[bc], feature_index[tx]), len(tx_umi_count[bc][tx]))
+            for bc in tx_umi_count
+            for tx in tx_umi_count[bc]
+        ),
+        shape=(len(barcode_list), len(feature_list)),
+        fill_value=0,
+        dtype=int).transpose()
```
Re-use the existing function rather than copy-pasting code.
```
    matrix, barcode_list, feature_list = io.isoquant_matrix(...)
```

jamestwebber replied:
actually, these functions (there should be one for each format) should take these as arguments

jamestwebber commented on src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py:
```diff
@@ -0,0 +1,123 @@
+
+import scanpy as sc
+
+def isoquant_matrix(
```
I'd rename this to something like "isoquant_to_h5ad" and put it in the `io` module

jamestwebber commented on src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py:
```diff
@@ -0,0 +1,123 @@
+    feature_list: Sequence[tuple[str, str]] = None,
+    out_isoform_seurat_path: str = "isoforms_seurat",
+    out_h5ad_path: str = "isoforms_h5ad",
+    gene_Ensembl_to_name: str = "gene_Ensembl_to_name.tsv"
```
this path is not going to work when the package is installed somewhere. It should just be a required argument

jamestwebber commented on src/mdl/isoscelles/isoquant_output_to_scanpy_seurat_input.py:
```diff
@@ -0,0 +1,123 @@
+        shutil.rmtree(out_h5ad_path)
+    os.mkdir(out_h5ad_path)
+
+    anndata = sc.read_10x_mtx(out_isoform_seurat_path)
+    anndata.write_h5ad(out_h5ad_path + "/isoform_feature_bc_matrix.h5ad")
```
This is going to be a really slow way to make an `h5ad` file: create a sparse matrix, writes it all out to text, reads it back _in_ and convert back to numbers, and then make the `h5ad` object. It's possible to create an `h5ad` directly from the data.

the `mtx` and `h5ad` formats can be built separately and should live in separate functions, as people might want one or the other and not necessarily both. They can both take the output of `isoquant_matrix` so that the file only needs to be read once.

jamestwebber reviewed (changes requested):
This needs a couple changes (some of my earlier comments are superfluous after all of this is put together):
 
 * this should be two functions, not one. One for making an `mtx` file and one for an `h5ad`
 * these functions should live in `mdl.isoscelles.io`, they don't need a new module
 * they should take advantage of the existing code rather than copying it. They can just take as inputs the results from `isoquant_matrix`
 * the file should be formatted with the existing code style

houlinyu committed:
Update isoquant_output_to_scanpy_seurat_input.py

made small changes, the input is compatible with the bam file of non-mapped reads. changed the feature list columns so it is now comparable with cell ranger output - convenient for downstream short versus long comparison using scanpy
